### PR TITLE
Download files via query data instead of html output

### PR DESCRIPTION
**Pull request by jbalboni**
## Description
This updates the Drupal asset download process to find urls in the GraphQL output, rather than the html output. This lets us catch them before they go into React applications or any other output. 

I also updated our build process to cache the assets, so that FE devs don't have to download them everytime they start a build.

## Testing done
Ran locally, verified that images were still showing up. Also verified that images in React widgets were using a local path, rather than drupal.

## Acceptance criteria
- [x] All images are still downloaded
- [x] Images are not downloaded when --pull-drupal flag is not used

## Definition of done
- [x] Events are logged appropriately
- [x] Documentation has been updated, if applicable
- [x] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

**Review by rianfowler — APPROVED** [review]
LGTM

File: src/site/stages/build/drupal/assets.js [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    Object.keys(current).forEach(key => {
+      let newValue = current;
+
+      if (typeof current[key] === 'string') {
+        newValue = replacer(current[key]);
+      } else {
+        newValue = replacePathInData(current[key], replacer);
+      }
+
+      if (newValue !== current[key]) {
```
[the remark this answers is not in the document]
Comment (jbalboni) [reply src/site/stages/build/drupal/assets.js]: I was trying not to mutate the object that's passed into this function, which may not be necessary, but seemed safer.

File: package.json [hunk 2]
```diff
@@ -164,6 +164,7 @@
     "puppeteer": "^1.9.0",
     "quick_check": "^0.7.0",
     "react-test-renderer": "^16.7.0",
+    "recursive-readdir": "^2.2.2",
```
Comment (ncksllvn) [inline package.json]: We should have done this long ago 💫

File: src/site/stages/build/drupal/assets.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+function convertAssetPath(drupalInstance, url) {
+  // After this path are other folders in the image paths,
+  // but it's hard to tell if we can strip them, so I'm leaving them alone
+  const withoutHost = url.replace(`${drupalInstance}/sites/default/files/`, '');
+  const path = withoutHost.split('?', 2)[0];
+
+  // This is sort of naive, but we'd like to have images in the img folder
+  if (
+    ['png', 'jpg', 'jpeg', 'gif', 'svg'].some(ext =>
+      path.toLowerCase().endsWith(ext),
```
Comment (ncksllvn) [inline src/site/stages/build/drupal/assets.js]: Not necessary, but [extname](https://nodejs.org/api/path.html#path_path_extname_path) could be nice here.

File: src/site/stages/build/plugins/download-drupal-assets.js [hunk 4]
```diff
@@ -1,135 +1,50 @@
...
-          file.contents = new Buffer(doc.html());
-        }
-      }
-    }
+    const assetsToDownload = Object.entries(files)
+      .filter(entry => entry[1].isDrupalAsset && !entry[1].contents)
+      .map(([key, value]) => ({
+        src: value.source,
+        dest: key,
+      }));
```
Comment (ncksllvn) [inline src/site/stages/build/plugins/download-drupal-assets.js]: This translation of the Drupal source to local destination took me a minute to follow, but this is really clever 🔥

**Review by ncksllvn — APPROVED** [review]
Looks great!

**Pull request merged**